Adding documentation for handleInstallPath, beforeRedirection, beforeInstallation, afterInstallation (#1448)
slackapi/node-slack-sdk: docs/_packages/oauth.md

### Initialize the installer

> This package exposes an `InstallProvider` class, which sets up the required configuration and exposes methods such as `generateInstallUrl`, `handleCallback`, `authorize` and `handleInstallPath` for use within your apps. At a minimum, `InstallProvider` takes a `clientId` and `clientSecret` (both which can be obtained under the **Basic Information** of your app configuration). `InstallProvider` also requires a `stateSecret`, which is used to encode the generated state, and later used to decode that same state to verify it wasn't tampered with during the OAuth flow. **Note**: This example is not ready for production because it only stores installations (tokens) in memory. Please go to the [storing installations in a database](#storing-installations-in-a-database) section to learn how to plug in your own database.

Reviewer: For new developers, the methods they should know would be `handleInstallPath`,  `handleCallback`, and `authorize`. We don't need to put emphasis on `generateInstallUrl` anymore.

```suggestion
This package exposes an `InstallProvider` class, which sets up the required configuration and exposes methods such as `handleInstallPath` (which calls `generateInstallUrl` internally), `handleCallback`, and `authorize` for use within your apps. At a minimum, `InstallProvider` takes a `clientId` and `clientSecret` (both which can be obtained under the **Basic Information** of your app configuration). `InstallProvider` also requires a `stateSecret`, which is used to encode the generated state, and later used to decode that same state to verify it wasn't tampered with during the OAuth flow. **Note**: This example is not ready for production because it only stores installations (tokens) in memory. Please go to the [storing installations in a database](#storing-installations-in-a-database) section to learn how to plug in your own database.
```

---

## Showing an Installation Page

You'll need an installation URL when you want to test your own installation in order to submit your app to the App Directory and in case you need additional authorizations (such as user tokens) from users inside a team where your app is already installed. These URLs are also commonly used on your own webpages as the link for an ["Add to Slack" button](https://api.slack.com/docs/slack-button).

There are two approaches the `InstallProvider` provides to show an installation page to the user prior to beginning the OAuth process:

> 1. You can let `InstallProvider` render the installation page at a URL/path of your choosing [using the `handleInstallPath()` method](#using-handleinstallpath). It will automatically display an "Add to Slack" button and encode any desired user or bot scopes and metadata you specify.
> 2. You can render the installation page yourself and [call `generateInstallUrl()`](#using-generateinstallurl) to generate a secure slack.com-based URL to initiate the OAuth process. You would use this generated URL in your own installation page's links/buttons/UI elements.

Reviewer: You can customize the behavior by passing `renderHtmlForInstallPath` function when initializing `InstallProvider`. Thus, there is no limitation on it in `handleInstallPath`. We can say that "For most use cases, we recommend going with `handleInstallPath()` with necessary customize options. If you need only the authorize URL generation, you can use `generateInstallUrl()`, which is called inside `handleInstallPath()`, in your code.

```suggestion
1. You can let `InstallProvider` render the installation page at a URL/path of your choosing [using the `handleInstallPath()` method](#using-handleinstallpath). It will automatically display an "Add to Slack" button and encode any desired user or bot scopes and metadata you specify.
2. You can customize the installation page by passing `renderHtmlForInstallPath` function to the `InstallProvider` constructor. Also, if your app supports Direct Install URL in the App Directory page, you can pass `directInstall: true` when initializing `InstallProvider`.
```

#### Using `handleInstallPath`

If you don't need to customize the installation page users will be shown, you can let this package render the installation page for you using the `handleInstallPath()` method.

```javascript
// Assume the installation page is located at /slack/install
app.get('/slack/install', async (req, res) => {
  await installer.handleInstallPath(req, res, {
    scopes: ['chat:write'],
    userScopes: ['channels:read'],
    metadata: 'some_metadata',
  });
});
```

The `handleInstallPath` method accepts an options object as its third argument which supports `scopes`, `metadata`, `userScopes`, `teamId` and `redirectUri` properties (check out the [source code for this interface](https://github.com/slackapi/node-slack-sdk/blob/main/packages/oauth/src/install-url-options.ts) for more details).

#### Using `generateInstallUrl`

> If you want to customize the installation page users will be shown, you may generate an installation URL dynamically and use the generated URL as part of the installation page displayed to the user.

Reviewer: As I mentioned above, we don't need to mention `generateInstallUrl` for this reason. We can suggest `renderHtmlForInstallPath` instead. We still can have `generateInstallUrl` section for backward-compatibility and advanced use cases to take 100% control of the endpoint handling.

Author: @seratch I've now changed this section to singularly focus on `handleInstallPath`, and moved any references to the old `generateInstallUrl` way under the details section. Hopefully that encourages developers down our recommended path (resolving this for now).

### Handling the OAuth redirect

> After the user approves the request to install your app (and grants access to the required permissions), Slack will redirect the user to your specified **redirect URL**. You can either set the redirect URL in the app’s **OAuth and Permissions** page or pass a `redirectUri` when calling `installProvider.generateInstallUrl`. Your HTTP server should handle requests to the redirect URL by calling the `installProvider.handleCallback()` method. The first two arguments (`req`, `res`) to `installProvider.handleCallback` are required. By default, if the installation is successful the user will be redirected back to your App Home in Slack (or redirected back to the last open workspace in your slack app for classic Slack apps). If the installation is not successful the user will be shown an error page.

Reviewer: ```suggestion
After the user approves the request to install your app (and grants access to the required permissions), Slack will redirect the user to your specified **Redirect URL**. You can either set the redirect URL in the app’s **OAuth and Permissions** page or pass a `redirectUri` when calling `installProvider.generateInstallUrl`. Your HTTP server should handle requests to the redirect URL by calling the `installProvider.handleCallback()` method. The first two arguments (`req`, `res`) to `installProvider.handleCallback` are required. By default, if the installation is successful the user will be redirected back to your App Home in Slack (or redirected back to the last open workspace in your slack app for classic Slack apps). If the installation is not successful the user will be shown an error page.
```
